src/FreeAge/common/type_stats_data.cpp

```diff
@@ -0,0 +1,330 @@
+  s.damage = GetBuildingDefaultDamage();
+  s.friendlyDamage = false;
+  s.populationSpace = 0;
+  s.dropOffPoint[static_cast<int>(ResourceType::Wood)] = false;
+  s.dropOffPoint[static_cast<int>(ResourceType::Food)] = false;
+  s.dropOffPoint[static_cast<int>(ResourceType::Gold)] = false;
+  s.dropOffPoint[static_cast<int>(ResourceType::Stone)] = false;
+  s.size = QSize(0, 0);
+  s.occupancy = QRect(0, 0, 0, 0);
+}
+
+inline void SetBuildingBasic(BuildingTypeStats& s, int maxHp, int meleeArmor, int pierceArmor, int size, float lineOfSight) {
+  s.maxHp = maxHp;
+  s.armor.SetValue(DamageType::Melee, meleeArmor);
+  s.armor.SetValue(DamageType::Pierce, pierceArmor);
+  s.armor.SetValue(DamageType::Building, 0);
+  s.armor.SetValue(DamageType::StandardBuilding, 0);
+  s.size = QSize(size, size);
+  s.occupancy = QRect(0, 0, size, size);
+  s.lineOfSight = lineOfSight;
+}
+
+inline void SetBuildingGarrison(BuildingTypeStats& s, GarrisonType garrisonType, int garrisonCapacity) {
+  s.garrisonType = garrisonType;
+  s.garrisonCapacity = garrisonCapacity;
+}
+
+inline void SetBuildingAttack(BuildingTypeStats& s, AttackType attackType, float fireRate, int pierceDamage, float accuracy, float maxRange, float projectileSpeed, float attackDelay = 0) {
+  s.attackType = attackType;
+  s.projectileSpeed = projectileSpeed;
+  s.fireRate = fireRate;
+  s.damage.SetValue(DamageType::Pierce, pierceDamage);
+  s.accuracy = accuracy;
+  s.maxRange = maxRange;
+  s.projectileSpeed = projectileSpeed;
+  s.attackDelay = attackDelay;
+}
+
+inline void SetBuildingBonusDamage(BuildingTypeStats& s, DamageType damageType, int value) {
+  s.damage.SetValue(damageType, value);
+}
+
+inline void SetBuildingCost(BuildingTypeStats& s, float creationTime, u32 wood, u32 food, u32 gold, u32 stone) {
+  s.creationTime = creationTime;
+  s.cost = ResourceAmount(wood, food, gold, stone);
+}
+
+// Vector helpers
+
+inline UnitTypeStats& InitUnit(std::vector<UnitTypeStats>& unitTypeStats, UnitType type) {
+  UnitTypeStats& s = unitTypeStats[static_cast<int>(type)];
+  SetUnitDefaults(s);
+  return s;
+}
+inline UnitTypeStats& CopyUnit(std::vector<UnitTypeStats>& unitTypeStats, UnitType type, UnitType source) {
+  UnitTypeStats& s = unitTypeStats[static_cast<int>(type)];
+  s = unitTypeStats[static_cast<int>(source)];
+  return s;
+}
+
+inline BuildingTypeStats& InitBuilding(std::vector<BuildingTypeStats>& buildingTypeStats, BuildingType type) {
+  BuildingTypeStats& s = buildingTypeStats[static_cast<int>(type)];
+  SetBuildingDefaults(s);
+  return s;
+}
+inline BuildingTypeStats& CopyBuilding(std::vector<BuildingTypeStats>& buildingTypeStats, BuildingType type, BuildingType source) {
+  BuildingTypeStats& s = buildingTypeStats[static_cast<int>(type)];
+  s = buildingTypeStats[static_cast<int>(source)];
+  return s;
+}
+
+// Data
+
+// NOTE: The following functions should be able to inline all the helper function calls and elimenate
+//       any redundant value sets.
+
```

Comment:
@puzzlepaint I added some of the units and buildings with this design of using helper inline functions. I use comments to make more clear what every value means, obviously this kind of format would be only for this file, which I think its ok because its not really code but just data. 
I know two people is basically impossible to agree in one format, but before I add all the current supported units: what do you think about how it looks?

The idea is to use this functions instead of just setting every value in order to avoid making mistakes and also compressing the size of the code.

Reply:
+1 for the inline comments, they seem helpful here (in case the IDE cannot show the parameter names inline).

For me it would seem more natural to make the `SetXY()` functions member functions of the Stats structs. The current style seems more like C than C++ ;) But it does not really matter much.

Reply:
I think to avoid having all the getters/setters we could just pass everywhere `const Stats` and only the load code and `Modification` code have access the no const (I think only Modification should be able to change the stats after the initial set up)

I know there is the _friend_ declaration to make setters work only in some places but I'm not sure how it works and if it's worth it...

Reply:
I am not sure whether I understand the concern here. With something like:
```c++
struct Stats {
  int attribute1;
  int attribute2;

  void SetAllUnitAttributes(int attribute1, int attribute2);
};
```
passing a `const Stats&` around would allow to read the attributes without the need for getters, and passing a non-const `Stats` around would allow to set the attributes without the need for a friend class/struct?

Reply:
So the `SetUnitBasic` would be "visible" everywhere but you could not use it because you have a "const" reference... I understand now. I have been working mostly with Java the last months, where something like this is not possible.

However all the Set functions will only ever be needed in the `type_stats_data.cpp` (and even be removed entirely if the data in the future are loaded from files).

By the way I thought `SetXY` meant having a setter for each field... this probably why you said XY instead of X.

Reply:
Sorry for not being clear enough. Yes, this would make the functions visible outside of `type_stats_data.cpp`. I am totally fine with both implementations, the more object-oriented one just "feels" a bit more natural because it keeps the code interacting with the structs closer to the structs themselves.

Reply:
After a bit of web searching: the setters could be private and the `LoadUnitTypeStats` could be declared as a friend, but then all the Set* would either have to be put on the hpp (which would pobably be included by a lot of files) or remove the `inline` keyword and put them to a cpp. I will leave it like this for now and in any case a future refactor would only be a couple of easy find-and-replace operations.